### PR TITLE
[17.05] Force uniqueness of preserved tags

### PR DESCRIPTION
**Pull request by dannon**
So, don't allow many inputs with a 'name:foo' tag to add a ton of 'name:foo' tags to a dataset; just use one.  Ping @mvdbeek

### CODE REVIEW

**Comment by martenson** [comment]
👍

File: lib/galaxy/tools/actions/__init__.py [hunk 1]
```diff
@@ -236,7 +236,7 @@ def execute(self, tool, trans, incoming={}, return_job=False, set_output_hid=Tru
                 incoming[ "%s|__identifier__" % name ] = identifier
 
             for tag in data.tags:
-                if tag.user_tname == 'name':
+                if tag.user_tname == 'name' and tag.value not in [t.value for t in preserved_tags]:
```
Comment (mvdbeek) [inline lib/galaxy/tools/actions/__init__.py]: This only addresses regular datasets (LDA/HDA), right? I'm still seeing duplicated tags with collections, but I'm not exactly sure how/where collections and tags interact.

Comment (dannon) [reply lib/galaxy/tools/actions/__init__.py]: Hrmm.  This should be the same (now filtered) list passed forward to the output_collections handling (https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/tools/actions/__init__.py#L409)

Got a minimal example with out-of-the-box tools I can try to see what you're seeing?

Comment (mvdbeek) [reply lib/galaxy/tools/actions/__init__.py]: Sure, upload two files, create a collection with one file each, tag both collections with the same tag. Then concatenate (with galaxy's concat1 tool) the two collections and you will have a duplicate tag.

Comment (dannon) [reply lib/galaxy/tools/actions/__init__.py]: Awesome, thanks.  Let me take a quick look.

Comment (nsoranzo) [reply lib/galaxy/tools/actions/__init__.py]: Probably they come from https://github.com/galaxyproject/galaxy/blob/dev/lib/galaxy/managers/collections.py#L106

Comment (martenson) [reply lib/galaxy/tools/actions/__init__.py]: It also propagates the non-hastag tags for collections.

Comment (martenson) [reply lib/galaxy/tools/actions/__init__.py]: As of the issue @mvdbeek reported it actually shows the tag only once in the tag ui (when expanded) but it shows it twice on the dataset (without expanding it)

Comment (dannon) [reply lib/galaxy/tools/actions/__init__.py]: Yeah, there's this second path of logic for this other type of collection output, patch inc in a sec.

**Comment by dannon** [comment]
(Putting this into WIP; it fixes one problem but there's another, and I'm going to slightly refactor this later -- might as well combine it all before we merge)

**Comment by martenson** [comment]
Given we released yesterday I am for quick fix and refactoring later in a different PR to minimize the number of people that install/encounter this bug.

**Comment by dannon** [comment]
This should work for now.  I'm still travelling, but will follow up with one more clientside fix that'll render only a single copy of each hashtag (just like the tag editor currently) as well.

**Comment by dannon** [comment]
I went ahead and added (2ac04bb) the display bit I mentioned previously.  This will now make existing multiple-duplicate-tagged datasets and collections display fine as well.

**Comment by martenson** [comment]
This works fine in my testing, someone please review & merge (or approve).

**Comment by mvdbeek** [comment]
Excellent, thanks a lot for fixing this @dannon (and the tags in general, they're awesome.)

**Pull request merged**